### PR TITLE
refactor(array): rename BytesWriter and make it recoverable

### PR DESCRIPTION
**Pull request by TennyZhuang**
Signed-off-by: TennyZhuang <zty0826@gmail.com>

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

1. Rename BytesWriter to StringWriter, the latter is better for us.
2. StringWriter doesn't hold the ownership of the builder, but recovers the dirty state on drop.
3. Remove unnecessary `Result`.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

**Comment by wangrunji0408** [comment]
An off-topic: why don't we rename all `Utf8` to `String` in the array module. 😇

**Comment by TennyZhuang** [comment]
> An off-topic: why don't we rename all `Utf8` to `String` in the array module. 😇

NTFS

Reason from @liurenjie1024 , we only support UTF8 collation, so we use UTF8 now.

File: src/common/src/array/utf8_array.rs [hunk 1]
```diff
@@ -331,21 +328,59 @@ mod tests {
...
-        Ok(())
+    #[test]
+    fn test_utf8_partial_writer_failed() {
+        let mut builder = Utf8ArrayBuilder::new(0);
+        // Write a record.
+        let _guard: WrittenGuard = {
+            let writer = builder.writer();
+            let mut partial_writer = writer.begin();
+            partial_writer.write_ref("Dia");
+            partial_writer.write_ref("na");
```
Comment (BugenZhao) [inline src/common/src/array/utf8_array.rs:355]: 🤤🤤🤤

**Review by BugenZhao — APPROVED** [review]
LGTM!

**Pull request merged**